### ISSUE TITLE
CSI volume does not attach normally after abnormal detach

### ISSUE THREAD

**Issue by carlosrbcunha** [kubernetes-sigs/azuredisk-csi-driver#584]
**What happened**:
After a docker/nomad clients reinstall (vm image update) which led terraform to shutdown the servers abruptly, CSI volume do not mount with error ```failed to setup alloc: pre-run hook "csi_hook" failed: node plugin returned an internal error, check the plugin allocation logs for more information: rpc error: code = Internal desc = could not format "/dev/disk/azure/scsi1/lun0"(lun: "0"), and mount it at "/csi/staging/nexus/rw-file-system-single-node-writer"```
**What you expected to happen**:
After a detach, the volume should attach again and allow usage.
**How to reproduce it**:
With a CSI disk attached to a VM, destroy it and replace the VM with another (can have the same name) and configure it to use CSI in the same manner.
**Anything else we need to know?**:
I have tried to access the disk and clean it (erase all partitions) and after that the disk is normally attached. Something is preventing the disk from being attached normally again.

csi-node logs
```
I1104 16:21:18.523533       1 utils.go:109] GRPC request: {}
I1104 16:21:18.524067       1 utils.go:114] GRPC response: {"ready":{"value":true}}
I1104 16:21:18.875709       1 azure_common_linux.go:175] azureDisk - found /dev/disk/azure/scsi1/lun0 by sdc under /dev/disk/azure/scsi1/
I1104 16:21:18.875737       1 nodeserver.go:121] NodeStageVolume: formatting /dev/disk/azure/scsi1/lun0 and mounting at /csi/staging/nexus/rw-file-system-single-node-writer with mount options([])
I1104 16:21:18.875751       1 mount_linux.go:405] Attempting to determine if disk "/dev/disk/azure/scsi1/lun0" is formatted using blkid with args: ([-p -s TYPE -s PTTYPE -o export /dev/disk/azure/scsi1/lun0])
I1104 16:21:18.896692       1 mount_linux.go:408] Output: "DEVNAME=/dev/disk/azure/scsi1/lun0\nPTTYPE=gpt\n", err: <nil>
I1104 16:21:18.896744       1 mount_linux.go:446] Disk /dev/disk/azure/scsi1/lun0 detected partition table type: gpt
W1104 16:21:18.896753       1 mount_linux.go:381] Configured to mount disk /dev/disk/azure/scsi1/lun0 as unknown data, probably partitions but current format is ext4, things might break
I1104 16:21:18.896758       1 mount_linux.go:298] Checking for issues with fsck on disk: /dev/disk/azure/scsi1/lun0
I1104 16:21:18.952326       1 utils.go:108] GRPC call: /csi.v1.Identity/Probe
I1104 16:21:18.952342       1 utils.go:109] GRPC request: {}
I1104 16:21:18.953651       1 utils.go:114] GRPC response: {"ready":{"value":true}}
I1104 16:21:18.963887       1 utils.go:108] GRPC call: /csi.v1.Node/NodeGetCapabilities
I1104 16:21:18.963901       1 utils.go:109] GRPC request: {}
I1104 16:21:18.964386       1 nodeserver.go:237] Using default NodeGetCapabilities
I1104 16:21:18.964392       1 utils.go:114] GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":3}}},{"Type":{"Rpc":{"type":2}}}]}
I1104 16:21:19.007592       1 mount_linux.go:311] `fsck` error fsck from util-linux 2.33.1
fsck.ext2: Bad magic number in super-block while trying to open /dev/sdc
/dev/sdc: 
The superblock could not be read or does not describe a valid ext2/ext3/ext4
filesystem.  If the device is valid and it really contains an ext2/ext3/ext4
filesystem (and not swap or ufs or something else), then the superblock
is corrupt, and you might try running e2fsck with an alternate superblock:
    e2fsck -b 8193 <device>
 or
    e2fsck -b 32768 <device>

Found a gpt partition table in /dev/sdc
I1104 16:21:19.007625       1 mount_linux.go:394] Attempting to mount disk /dev/disk/azure/scsi1/lun0 in ext4 format at /csi/staging/nexus/rw-file-system-single-node-writer
I1104 16:21:19.007647       1 mount_linux.go:146] Mounting cmd (mount) with arguments (-t ext4 -o defaults /dev/disk/azure/scsi1/lun0 /csi/staging/nexus/rw-file-system-single-node-writer)
E1104 16:21:19.023425       1 mount_linux.go:150] Mount failed: exit status 32
Mounting command: mount
Mounting arguments: -t ext4 -o defaults /dev/disk/azure/scsi1/lun0 /csi/staging/nexus/rw-file-system-single-node-writer
Output: mount: /csi/staging/nexus/rw-file-system-single-node-writer: wrong fs type, bad option, bad superblock on /dev/sdc, missing codepage or helper program, or other error.

E1104 16:21:19.023461       1 utils.go:112] GRPC error: rpc error: code = Internal desc = could not format "/dev/disk/azure/scsi1/lun0"(lun: "0"), and mount it at "/csi/staging/nexus/rw-file-system-single-node-writer"
I1104 16:21:45.251187       1 utils.go:108] GRPC call: /csi.v1.Node/NodeUnpublishVolume
I1104 16:21:45.251207       1 utils.go:109] GRPC request: {"target_path":"/csi/per-alloc/865651fa-26ef-d8c6-28c5-c147bc974fa1/nexus/rw-file-system-single-node-writer","volume_id":"/subscriptions/11111111-2222-3333-4444-555555555555/resourceGroups/test-rg/providers/Microsoft.Compute/disks/nexus-disk"}
I1104 16:21:45.251759       1 nodeserver.go:225] NodeUnpublishVolume: unmounting volume /subscriptions/11111111-2222-3333-4444-555555555555/resourceGroups/test-rg/providers/Microsoft.Compute/disks/nexus-disk on /csi/per-alloc/865651fa-26ef-d8c6-28c5-c147bc974fa1/nexus/rw-file-system-single-node-writer
W1104 16:21:45.251787       1 mount_helper_common.go:33] Warning: Unmount skipped because path does not exist: /csi/per-alloc/865651fa-26ef-d8c6-28c5-c147bc974fa1/nexus/rw-file-system-single-node-writer
I1104 16:21:45.251793       1 nodeserver.go:230] NodeUnpublishVolume: unmount volume /subscriptions/11111111-2222-3333-4444-555555555555/resourceGroups/test-rg/providers/Microsoft.Compute/disks/nexus-disk on /csi/per-alloc/865651fa-26ef-d8c6-28c5-c147bc974fa1/nexus/rw-file-system-single-node-writer successfully
I1104 16:21:45.251799       1 utils.go:114] GRPC response: {}
I1104 16:21:45.252735       1 utils.go:108] GRPC call: /csi.v1.Node/NodeUnstageVolume
I1104 16:21:45.252745       1 utils.go:109] GRPC request: {"staging_target_path":"/csi/staging/nexus/rw-file-system-single-node-writer","volume_id":"/subscriptions/11111111-2222-3333-4444-555555555555/resourceGroups/test-rg/providers/Microsoft.Compute/disks/nexus-disk"}
I1104 16:21:45.253236       1 nodeserver.go:144] NodeUnstageVolume: unmounting /csi/staging/nexus/rw-file-system-single-node-writer
W1104 16:21:45.253258       1 mount_helper_common.go:65] Warning: "/csi/staging/nexus/rw-file-system-single-node-writer" is not a mountpoint, deleting
I1104 16:21:45.253303       1 nodeserver.go:149] NodeUnstageVolume: unmount /csi/staging/nexus/rw-file-system-single-node-writer successfully
I1104 16:21:45.253309       1 utils.go:114] GRPC response: {}
I1104 16:21:48.525838       1 utils.go:108] GRPC call: /csi.v1.Identity/Probe
I1104 16:21:48.525856       1 utils.go:109] GRPC request: {}
I1104 16:21:48.526309       1 utils.go:114] GRPC response: {"ready":{"value":true}}
I1104 16:21:48.967023       1 utils.go:108] GRPC call: /csi.v1.Identity/Probe
```

**Environment**:
- CSI Driver version: Experience this error in v0.9 and v0.10
- Nomad version : Nomad v0.12.4 (8efaee4ba5e9727ab323aaba2ac91c2d7b572d84)
- OS (e.g. from /etc/os-release): ubuntu 18.04.5 LTS
- Kernel (e.g. `uname -a`): 5.4.0-1031-azure #32~18.04.1-Ubuntu SMP Tue Oct 6 10:03:22 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
- Install tools:
- Others:

**Comment by andyzhangx** [comment]
it looks like your disk partition is broken, right?  So actually the disk could be attached, while it cannot mount the file system in that disk, right?

**Comment by carlosrbcunha** [comment]
Sorry for the late response Andy but , Yes, the partition is broken.
My question is more to why did this happened  ?? and how to prevent this.
Many times in our infra we start to get errors restarting tasks and we spot that one of the controller tasks is not functioning properly. Its running but we cannot see it registering in the nomad plugin. Restarting the nodes and controllers solves the issue but its recurring. 
How can we check effectively If a node or a controller is not properly working or connected to Azure ?

**Comment by andyzhangx** [comment]
by default, azure disk only has one partition, while would you need to erase all partitions?
About healthy check, there is `/ListVolumes` in controller logs and `/csi.v1.Node/NodeGetInfo` in node logs, that could check whether those components are connected to Azure. If there is node or controller failure, pls provide:
1. pod events to locate which container is not running well
2. get container logs

Could follow with this guide: https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/master/docs/csi-debug.md#csi-driver-debug-tips

**Comment by andyzhangx** [comment]
BTW, even driver does not function well, it should not corrupt disk partitions

**Comment by carlosrbcunha** [comment]
I have to agree with you on that. It should not corrupt disk partitions and allow them to seamlessly mount. The only thing that can justify this is the disk be in use by the workload and, being forcibly detached by the controller because of unknown status to try and attach to another node and try again. I saw this behaviour when having problems mounting the disk and with partition format issues. The controller tried to move the disk around from machine to machine to complete the attach successfully.
I will try to give you more data on what is happening.
Thanks for your help.

**Comment by andyzhangx** [comment]
> I have to agree with you on that. It should not corrupt disk partitions and allow them to seamlessly mount. The only thing that can justify this is the disk be in use by the workload and, being forcibly detached by the controller because of unknown status to try and attach to another node and try again. I saw this behaviour when having problems mounting the disk and with partition format issues. The controller tried to move the disk around from machine to machine to complete the attach successfully.
> I will try to give you more data on what is happening.
> Thanks for your help.

@carlosrbcunha how do you use the disk? With this disk driver, it can only have one partition, not sure why there are other partitions in the disk.

**Comment by carlosrbcunha** [comment]
@andyzhangx It only had one partition. I wrote partitions by mistake.

Basically the problem was that something happened and the workload stopped and when we started the node and controller and started the workload again. The disk attached but when mounting there was a problem identifying the partition and the disk cannot be mounted. I had to manually attach the disk to a server and erase all the disk data before the plugin could start using the disk again. 
Until now I do not know what happened.
